### PR TITLE
Add client side ExperimentDefinition

### PR DESCRIPTION
**Pull request by panahiparham**
Addresses #7

### CODE REVIEW

File: ml_experiment/ExperimentDefinition.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                return [
+                    {**c, 'seed': seed}
+                    for c in _c
+                    for seed in product_seeds
+                ]
+            else:
+                return _c
+
+
+    def get_results_path(self) -> str:
```
Comment (andnp) [inline ml_experiment/ExperimentDefinition.py]: We will want some single source-of-truth for this function. Most of it can be refactored out into a utility, I think:
```python
def get_experiment_name():
  import __main__
  return __main__.__file__.split('/')[-2]

def get_results_path(base_path: str, name: str):
  return os.path.join(
    base_path,
    name,
    get_experiment_name(),
  )
```

Then both sides can reference this (the def'n creator side and the client side).

File: ml_experiment/ExperimentDefinition.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            cur = con.cursor()
+            # ensure config_id exists
+            assert config_id in self.table.get_configuration_ids(cur),\
+            f'config_id not in table {self.table.get_table_name()}'
+
+            return self.table.get_configuration(cur, config_id)
+
+
+
+    def get_configs(self, config_ids: list[int], product_seeds: list[int] | None = None) -> list[dict[str, Any]]:
```
Comment (andnp) [inline ml_experiment/ExperimentDefinition.py:27]: Overall, I think doing the loop "externally" here is just fine! No need to push this loop into the db just yet, I suspect the performance difference will be in the ms range at worst.

File: ml_experiment/ExperimentDefinition.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        save_path = self.get_results_path()
+        db_path = os.path.join(save_path, 'metadata.db')
+
+        with sqlite3.connect(db_path) as con:
+            cur = con.cursor()
+            # ensure config_ids exist
+            assert all(
```
Comment (andnp) [inline ml_experiment/ExperimentDefinition.py]: It might be worth pushing the assertion down into the:
```python
self.table.get_configuration(cur, config_id)
```
code instead. This way, we can be more specific with our error message:
```python
raise ValueError(f"config_id <{config_id}> is not in table <{self.get_table_name()}>")
```

File: ml_experiment/metadata/MetadataTable.py [hunk 4]
```diff
@@ -70,6 +70,17 @@ def get_configuration_id(self, cur: sqlite3.Cursor, configuration: Dict[str, Val
...
+    def get_configuration(self, cur: sqlite3.Cursor, config_id: int) -> Dict[str, ValueType]:
+        table_name = self.get_table_name()
+        cols = list(self.get_columns(cur))
+        col_str = ', '.join(cols)
+        res = (
+            cur.execute(f"SELECT {col_str} FROM '{table_name}' WHERE id=?", (config_id,))
+            .fetchone()
+        )
+
+        return {cols[i]: res[i] for i in range(len(cols))}
```
Comment (andnp) [inline ml_experiment/metadata/MetadataTable.py]: Might be a good spot for `zip`:
```python
return {k: v for k, v in zip(cols, res, strict=True)}
```
and that refactor makes it clear we are creating a dict out of a list of tuples, which allows an even simpler refactor:
```python
return dict(zip(cols, res, strict=True))
```

Comment (andnp) [reply ml_experiment/metadata/MetadataTable.py]: Though actually seeing both side-by-side, I can't decide which refactor I like better.

File: tests/test_ExperimentDefinition.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import os
+
+from ml_experiment.DefinitionPart import DefinitionPart
+from ml_experiment.ExperimentDefinition import ExperimentDefinition
+
+
+def test_ExperimentDefinition():
```
Comment (andnp) [inline tests/test_ExperimentDefinition.py]: This is a great spot for the `tmp_path` fixture, which produces (and cleans up) a temporary file path for storing test results and guaranteeing there is no overlapping global state.
```python
def test_ExperimentDefinition(tmp_path):
  ...
  part = stubbed_DefinitionPart(exp_name, part_name, base=str(tmp_path))
  ...
```

File: tests/test_ExperimentDefinition.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    config_ids = [1, 2, 3]
+    seeds = [1, 2]
+
+    exp = stubbed_ExperimentDefinition(exp_name, part_name, version)
+
+    config = exp.get_config(0)
+    assert config == {'alpha': 0.125, 'beta': 0.5, 'id': 0}
+
+    configs = exp.get_configs(config_ids)
+    assert configs == [
```
Comment (andnp) [inline tests/test_ExperimentDefinition.py:30]: This does make me wonder if we should consider making some guarantees about ordering. I'm leaning towards "yes", since that's also what PyExpUtils does.

We can make that an issue, then do the guaranteed ordering in another PR.

**Review by andnp — CHANGES_REQUESTED** [review]

File: ml_experiment/ExperimentDefinition.py (hunk 1 above)
Comment (panahiparham) [reply ml_experiment/ExperimentDefinition.py:56]: Agreed! I am adding this change to this pr.

File: tests/test_ExperimentDefinition.py (hunk 5 above)
Comment (panahiparham) [reply tests/test_ExperimentDefinition.py]: Sorcery!

File: ml_experiment/ExperimentDefinition.py (hunk 1 above)
Comment (panahiparham) [reply ml_experiment/ExperimentDefinition.py]: Agreed! But this change breaks `stubbed_DefinitionPart` in `tests/test_ExperimentDefinition.py` as this class stubs the `get_results_path` method. For now I added a property that points to the 'single source of truth` function that we can override in tests.

```python
from ml_experiment._utils.path import get_results_path

class ExperimentDefinition:
    def __init__(self, ...):
        self.get_results_path = get_results_path
        ...

```

File: ml_experiment/metadata/MetadataTable.py (hunk 4 above)
Comment (panahiparham) [reply ml_experiment/metadata/MetadataTable.py]: I like `dict(zip(A,B))`. Should this refactor be part of this pr?

File: ml_experiment/ExperimentDefinition.py (hunk 3 above)
Comment (panahiparham) [reply ml_experiment/ExperimentDefinition.py]: Makes sense!

File: ml_experiment/metadata/MetadataTable.py (hunk 4 above)
Comment (andnp) [reply ml_experiment/metadata/MetadataTable.py]: Yup, I would make it part of this pr since this is the place it is being added. I typically only punt tech debt clean up when it is in code that is technically outside the pr, or when it adds a significant amount of work to the pr.

**Review by andnp — APPROVED** [review]
Looks great! I would go ahead and do the `dict(zip)` change and whatever git cleanup you'd like, then it's good to merge!

File: tests/test_ExperimentDefinition.py (hunk 6 above)
Comment (panahiparham) [reply tests/test_ExperimentDefinition.py:30]: Created #11 for this feature!

**Comment by panahiparham** [comment]
After clean up this pr has 5 commits and resolves issue #7 by adding an `experiment_definition` class that handles the client side interaction with the experiment after it is created and stored in a metadata database.

Commit 1 and 2 implement the `experiment_definition` class and its corresponding test. Commit 3 and 4 implement a utility for getting experiment path and commit 5 adds an init file to allow pytest to import this package.
